aten/src/ATen/Quantizer.cpp

```diff
@@ -0,0 +1,134 @@
+#include <ATen/ATen.h>
+#include <ATen/Quantizer.h>
+#include <ATen/QTensorImpl.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/native/TensorFactories.h>
+
+namespace at {
+
+std::shared_ptr<Quantizer> make_per_layer_affine_quantizer(double scale, int64_t zero_point) {
+  return std::make_shared<PerLayerAffineQuantizer>(static_cast<float>(scale), static_cast<int32_t>(zero_point));
+}
+
+// This is an internal utility function for getting at the QTensorImpl,
+// You should only use this for writing low level
+// setters/getters for QTensorImpl fields; otherwise, you should use
+// the low level setters/getters that were implemented using this.
+// This may be called repeatedly, so make sure it's pretty cheap.
+inline QTensorImpl* get_qtensorimpl(const QTensor& self) {
+  // TODO: remove this when Variable and Tensor are merged
+  AT_ASSERTM(!self.is_variable(), "_internal_get_QTensorImpl: should not be a variable");
+  // TODO: uncomment
+  // AT_ASSERTM(self.is_quantized(), "_internal_get_QTensorImpl: not a quantized tensor");
+  return static_cast<QTensorImpl*>(self.unsafeGetTensorImpl());
+}
+
+inline QTensor new_qtensor(
+    IntArrayRef sizes, const TensorOptions& options, float scale, int32_t zero_point) {
+  AT_ASSERT(options.device().is_cpu());
+
+  native::check_size_nonnegative(sizes);
+  auto* allocator = at::getCPUAllocator();
+  int64_t nelements = at::prod_intlist(sizes);
+  // TODO get from options
+  auto dtype = options.dtype();
+  // TODO: check wheather dtype is compatible with quantization
+  auto storage_impl = c10::make_intrusive<StorageImpl>(
+    dtype,
+    nelements,
+    allocator->allocate(nelements * dtype.itemsize()),
+    allocator,
+    /*resizable=*/true);
+  auto quantizer = make_per_layer_affine_quantizer(scale, zero_point);
+  auto tensor = detail::make_tensor<QTensorImpl>(
+      storage_impl, at::CPUTensorId(), false, quantizer);
+  // Default TensorImpl has size [0]
+  if (sizes.size() != 1 || sizes[0] != 0) {
```

Comment:
Err... what? If you only have sizes, it's always contiguous no matter what.

Reply:
This is copied from empty_cpu.
Why do we have that there?

Reply:
hm, it looks like some microoptimization from https://github.com/pytorch/pytorch/commit/db5aeafa6074e480a3ebd312963a7cc876d5a129 - probably should be fine to drop as we rarely create zero-sized tensors